replace math with joml

fayer3 opened the pull request:
this is built upon #284 so that needfs to be merged first

replaces old math classes with joml
moves the MathUtils class into common, and renames Utils to ClientUtils
adds debug render options to show device positions

fayer3 committed:
use joml instead of custom math classes.
use Vector3f for direction vectors, and Vec3 for positions
fix screenshot camera position in room vr data
remove creation of temp VRData objects just to convert values

fayer3 committed:
move MathUtils to the common package
rename the Utils in the client package to ClientUtils

fayer3 committed:
fix telescope aim

fayer3 committed:
add debug axis rendering, and debug server particles

fayer3 committed:
fix main menu gui rotation in seated constantly rotating

fayer3 committed:
replace degree/radian conversions that converted float stuff with a float multiplication

fayer3 committed:
make gunAngle a float

fayer3 committed:
fix seated snapping to head

fayer3 committed:
replace Math.PI in float contexts with Mth.PI and use the HALF and TWO variants
make rotinfo.bodyYaw return a float
replace some missed deg/rad conversions with the Mth constants

fayer3 committed:
fix default 3rd cam rotation

fayer3 committed:
remove some debug code

fayer3 committed:
change debug render helper statics to CAPS

fayer3 committed:
use Vec3.ZERO when applicable, and use Vector3f for the jump latch

fayer3 committed:
use DebugRenderHelper for the VRWidget debug axes, remove circle renderer

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/VRData.java:
```diff
@@ -96,35 +100,33 @@ public VRData(Vec3 origin, float walkMul, float worldScale, float rotation) {
             this.c2 = new VRDevicePose(this,
                 mcVR.getAimRotation(2),
-                mcVR.getAimSource(2).add(scaleOffset),
+                mcVR.getAimSource(2).add(scaleOffset, new Vector3f()),
```
Might as well put this one with the other controllers

fayer3 replied:
you mean move it above telescope and screenshot camera?

ferriarnus replied:
Put it next to `mainAimSource ` and `offAimSource`, as a third "controller".

fayer3 replied:
after the hand data, or between c and h?

ferriarnus replied:
I was thinking line 75, when c0 and c1 are done. Might also put c2 their to get all controllers next to each other? Though it doesn't really matter, was just wondering if it would make it more clear.

fayer3 replied:
well c2 isnt a fixed thing, there is also the option to have the camera at a fixed location, so I would need to move that whole if/else case there

ferriarnus replied:
Let's just leave it like this then, it's fine.

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/screenhandlers/GuiHandler.java:
```diff
@@ -216,52 +215,49 @@ public static void processGui() {
-            guiPos.setX((float) guiPos_room.x);
-            guiPos.setY((float) guiPos_room.y);
-            guiPos.setZ((float) guiPos_room.z);
+            float guiHalfWidth = 0.5F;
```
should this just be in-lined?

fayer3 replied:
I can do that if you want, dont really see why it is needed to not be inlined

ferriarnus replied:
I wasn't sure it meant much, but we can keep it like this for clarity.

fayer3 replied:
don't think there is much benefit from having them there. will inline them

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/screenhandlers/GuiHandler.java:
```diff
@@ -586,28 +579,28 @@ public static Vec3 applyGUIModelView(RenderPass currentPass, PoseStack pMatrixSt
+            yawAvg = Mth.DEG_TO_RAD * yawAvg;
 
-            Vec3 dir = new Vec3(-Math.sin(yawAvg), 0.0D, Math.cos(yawAvg));
+            Vector3f dir = new Vector3f((float) -Math.sin(yawAvg), 0.0F, (float) Math.cos(yawAvg));
```
Actually, now I see this, Minecrafts Mth lib has a float sin and cos function. Internally that uses this one, but then the cast isn't needed.

fayer3 replied:
the Minecraft Mth, uses a precomputed lookup table. I can change those though, if you think that is better

ferriarnus replied:
It should be fine I think? They probably did it in a way you won't notice the difference since we cast to float.

fayer3 replied:
so should I replace all Math.sin/cos calls (that are used for floats) with the minecraft Mth lib?

ferriarnus replied:
I think that would be fine to do yeah

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/screenhandlers/KeyboardHandler.java:
```diff
@@ -90,46 +91,47 @@ public static void orientOverlay(boolean guiRelative) {
         } else {
             // copy from GuiHandler.onScreenChanged for static screens
-            Vec3 offset = new Vec3(0.0D, -0.5D, -2.0D);
+            Vector3f offset = new Vector3f(0.0F, 0.0F, -2.0F);
```
did a -0.5 got lost?

fayer3 replied:
hm, need to check if I removed that on purpose

fayer3 replied:
okay yeah that was a mistake from me

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/CameraTracker.java:
```diff
@@ -34,8 +35,8 @@ public CameraTracker(Minecraft mc, ClientDataHolderVR dh) {
             return false;
-        } else if (this.dh.vrSettings.seated) {
-            return false;
+        /*} else if (this.dh.vrSettings.seated) {
```
Why was this special cased in seated?

fayer3 replied:
oh this is probably some leftover debug stuff, forgot to readd that again.
I assume it is hidden in seated, since it might be annoying?

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/ClimbTracker.java:
```diff
@@ -588,22 +594,22 @@ public void doProcess(LocalPlayer player) {
 
             if (movement.length() > limit) {
-                movement = movement.scale(limit / movement.length());
+                 movement.mul(limit / movement.length());
```
some weird space

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/ClimbTracker.java:
```diff
@@ -588,22 +594,22 @@ public void doProcess(LocalPlayer player) {
 
             if (player.hasEffect(MobEffects.JUMP)) {
-                movement = movement.scale(player.getEffect(MobEffects.JUMP).getAmplifier() + 1.5D);
+                 movement.mul(player.getEffect(MobEffects.JUMP).getAmplifier() + 1.5F);
```
some weird space

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/JumpTracker.java:
```diff
@@ -119,13 +123,14 @@ public void doProcess(LocalPlayer player) {
-            Vec3 rPos = this.dh.vrPlayer.vrdata_room_pre.getController(0).getPosition();
-            Vec3 lPos = this.dh.vrPlayer.vrdata_room_pre.getController(1).getPosition();
-            Vec3 now = rPos.add(lPos).scale(0.5D);
+            Vector3f now = this.dh.vrPlayer.vrdata_room_pre.getController(0).getPositionF().lerp(
```
Why use lerp now?

fayer3 replied:
a lerp can be faster, since it can use fma, but appart from that no real reason.

ferriarnus replied:
Ok, seems fine then

ferriarnus reviewed (commented):
Seems fine, just some questions

fayer3 committed:
use Mth.sin/cos when applicable
fix keyboard offset
fix camera widget shown in seated
fix some formatting errors

fayer3 committed:
inline gui half size
